File: 1-two-sum/1-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int> nums, int target) {
        int n = nums.size();
        unordered_set<int> s;
        vector<int> result;
        
        for (int i = 0; i < n; i++) {
            int x = target - nums[i];
            int index;
            //find index
            if (find(nums.begin(), nums.end(), x) != nums.end()) {
                index = find(nums.begin(), nums.end(), x) - nums.begin();
            }
            
            if (s.find(x) != s.end()) {
                result.push_back(index);
                result.push_back(i);
            }
            s.insert(nums[i]);
        }
        return result;
        
    }
};

File: 1-two-sum/1-two-sum.java
class Solution {
    public static int findIndex(Map<Integer,Integer> map, int target) {
        int index = 0;
        for (Map.Entry<Integer, Integer> mp: map.entrySet()) {
            if (mp.getKey() == target) {
                System.out.println("index: " + mp.getValue());
                index = mp.getValue();
            }
        }
        return index;
    }
    
    public int[] twoSum(int[] arr, int target) {
        // Brute Force
        
        // List<Integer> list = new ArrayList<>();
        // for (int i = 0; i < arr.length; i++) {
        //     for (int j = i+1; j < arr.length; j++) {
        //         if (arr[i] + arr[j] == target) {
        //             list.add(i);
        //             list.add(j);
        //         }
        //     }
        // }
        // System.out.println("List: " + list);
        // int[] result = new int[list.size()];
        // int i = 0;
        // for (int index : list) {
        //     result[i++] = index;
        // }
        // return result;
        
        
        // Hashing
        List<Integer> list = new ArrayList<>();
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < arr.length; i++) {
            if (map.containsKey(target - arr[i])) {
                int index = findIndex(map, target - arr[i]);
                System.out.println("INDEX: " + index);
                list.add(index);
                list.add(i);
//                map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);
            } else {
                map.put(arr[i], i);
            }
        }
        System.out.println("list: " + list);
        int[] result = new int[list.size()];
        int i = 0;
        for (int index : list) {
            result[i++] = index;
        }
        return result;
    }
}